utils/global-state-update-gen/src/auction_utils.rs

```diff
@@ -0,0 +1,115 @@
+            let bonded_amount = U512::from_dec_str(bonded_str).unwrap();
+            let seigniorage_recipient =
+                SeigniorageRecipient::new(bonded_amount, Default::default(), Default::default());
+            let _ = era_validators.insert(key, seigniorage_recipient);
+        }
+        let _ = new_snapshot.insert(era_id, era_validators);
+    }
+
+    new_snapshot
+}
+
+/// Returns the set of public keys that have bids larger than the smallest bid among the new
+/// validators.
+pub fn find_large_bids(
+    builder: &mut LmdbWasmTestBuilder,
+    new_snapshot: &SeigniorageRecipientsSnapshot,
+) -> BTreeSet<PublicKey> {
+    let max_bid = new_snapshot
+        .values()
+        .next()
+        .unwrap()
+        .values()
+        .map(SeigniorageRecipient::stake)
+        .min()
+        .unwrap();
+    builder
+        .get_bids()
+        .into_iter()
+        .filter(|(_pkey, bid)| bid.staked_amount() >= max_bid)
+        .map(|(pkey, _bid)| pkey)
+        .collect()
+}
+
+/// Generates a set of writes necessary to "fix" the bids, ie.:
+/// - set the bids of the new validators to their desired stakes,
+/// - remove the bids of the old validators that are no longer validators,
+/// - remove all the bids that are larger than the smallest bid among the new validators
+/// (necessary, because such bidders would outbid the validators decided by the social consensus).
+pub fn fix_bids(
+    builder: &mut LmdbWasmTestBuilder,
+    validators_diff: &ValidatorsDiff,
+    new_snapshot: &SeigniorageRecipientsSnapshot,
+) -> BTreeMap<Key, StoredValue> {
+    let large_bids = find_large_bids(builder, new_snapshot);
+    let to_unbid = validators_diff.removed.union(&large_bids);
+
+    validators_diff
+        .added
+        .iter()
+        .map(|pkey| {
+            let amount = *new_snapshot
+                .values()
+                .next()
+                .unwrap()
+                .get(pkey)
+                .unwrap()
+                .stake();
+            let account_hash = pkey.to_account_hash();
+            let account = builder.get_account(account_hash).unwrap();
+            (
+                Key::Bid(account_hash),
+                Bid::unlocked(*pkey, account.main_purse(), amount, Default::default()).into(),
+            )
+        })
+        .chain(to_unbid.into_iter().map(|pkey| {
+            let account_hash = pkey.to_account_hash();
+            let account = builder.get_account(account_hash).unwrap();
+            (
+                Key::Bid(account_hash),
+                Bid::empty(*pkey, account.main_purse()).into(),
+            )
+        }))
+        .collect()
+}
+
+/// Removes pending withdraws of all the old validators that cease to be validators.
+pub fn fix_withdraws(
+    builder: &mut LmdbWasmTestBuilder,
+    validators_diff: &ValidatorsDiff,
+) -> BTreeMap<Key, StoredValue> {
+    let withdraws = builder.get_withdraws();
+    let withdraw_keys: BTreeSet<_> = withdraws.keys().collect();
```

Comment:
We don't need `withdraws` anywhere so I would collapse these two to a single line
```rust
let withdraw_keys = builder.get_withdraws().keys().collect();
```

Reply:
It's actually needed. Without assigning this to a variable, the compiler wants to drop it at the line creating `withdraw_keys` - but since we're collecting references to the keys, that would be invalid. So this just serves to keep the lifetimes correct.